import { Link } from '@brillout/docpress'

> To block unauthorized access see <Link href="/permissions" /> instead, and for handling invalid `<form>` values see <Link href="/form-validation" /> instead.

## Bugs

If a telefunction has a bug:

```js
// hello.telefunc.js
// Environment: Node.js

export { hello }

function hello(name) {
  // This telefunction has a bug: it should be `name` instead of `namee`
  return 'Hello ' + namee;
}
```

Then the telefunction call throws an error:

```html
<!-- index.html -->
<!-- Environment: Browser -->

<html>
  <body>
    <script type="module">
      import { hello } from './hello.telefunc.js'

      try {
        await hello('Eva')
        console.log("I'm never printed")
      } catch(err) {
        console.log(err.message) // Prints 'Server Error'
      }
    </script>
  </body>
</html>
```

This means we can treat bugs occuring in our telefunctions just like any other bug.

> Just like with other bugs, we usually use our UI Framwork (React, Vue, ...) to handle bugs occuring during a telefunction call.


## Network Errors

If the user's browser can't connect to our server:

```html
<!-- index.html -->
<!-- Environment: Browser -->

<html>
  <body>
    <script type="module">
      import { hello } from './hello.telefunc.js'

      try {
        await hello('Eva')
      } catch(err) {
        if (err.isConnectionError) {
          // There is a network problem:
          //  - the user isn't connected to the internet, or
          //  - our server is down.
          console.log(err.message) // Prints 'No Server Connection'
        }
      }
    </script>
  </body>
</html>
```


## Error tracking

To track bugs, we use `onBug()`:

```js
import { onBug } from 'telefunc'

onBug(err => {
  // ...
})
```

Allowing us, for example, to install the tracker code of a tracking service ([Sentry](https://sentry.io/), [Bugsnag](https://www.bugsnag.com/), [Rollbar](https://rollbar.com/), ...).

`onBug()` is called:
 - When one of your telefunction throws an error that is not `Abort()`. (In other words, your telefunction has a bug.)
 - When Telefunc throws an error. (In other words, Telefunc has a bug).

> `throw Abort()` does *not* trigger `onBug()`. It's expected that `throw Abort()` may be called (when some third-party erroneously calls a telefunction). In other words, `throw Abort()` isn't a bug.
